File: lecture005/binarysearch.java
import java.util.Scanner;
public class binarySearch {
	public static Scanner scn = new Scanner(System.in);

	public static void main(String[] args) {
		int[] arr = new int[10];
		input(arr);
		int var = scn.nextInt();
		System.out.println(binSearch(arr, var));
		System.out.println(upperBinSearch(arr, var));
		System.out.println(lowerBinSearch(arr, var));
	}
    public static void input(int[] arr) {
		for (int i = 0; i < arr.length; i++) {
			arr[i] = scn.nextInt();
		}
	}

	//  binary search
	public static int binSearch(int[] arr, int var) {
		int upperBound = 0;					//  starting index
		int lowerBound = arr.length;		//  ending index
		while(lowerBound > upperBound) {
			int middle = (upperBound + lowerBound)/2;
			if (arr[middle] == var) {
				return middle;
			}
			else if (arr[middle] < var) {
				upperBound = middle + 1;
			}
			else if (arr[middle] > var) {
				lowerBound = middle - 1;
			}
		}

		if(lowerBound == -1){
			return upperBound;
		}
		else if(upperBound == arr.length) {
			return lowerBound;
		}
		else if (var - arr[lowerBound] <= arr[upperBound] - var){
			return lowerBound;
		}
		return -1;		//  element doesn't exist
	}

	//  upper bound binary search
	public static int upperBinSearch(int[] arr, int var) {
		int upperBound = 0;					//  starting index
		int lowerBound = arr.length - 1;		//  ending index
		while(lowerBound > upperBound) {
			int middle = (upperBound + lowerBound)/2;
			if (arr[middle] == var) {
				if (middle + 1 < arr.length && arr[middle + 1] == arr[middle]) {
					upperBound = middle + 1;
				}
				else{
					return middle;
				}
			}
			else if (arr[middle] < var) {
				upperBound = middle + 1;
			}
			else if (arr[middle] > var) {
				lowerBound = middle - 1;
			}
		}
		return -1;		//  element doesn't exist
	}

	// lower bound binary search
	public static int lowerBinSearch(int[] arr, int var) {
		int upperBound = 0; // starting index
		int lowerBound = arr.length - 1; // ending index
		while (lowerBound >= upperBound) {
			int middle = (upperBound + lowerBound) / 2;
			if (arr[middle] == var) {
				if (middle - 1 >= 0 && arr[middle - 1] == arr[middle]) {
					lowerBound = middle - 1;
				}
				else {
					return middle;
				}
			} else if (arr[middle] < var) {
				upperBound = middle + 1;
			} else if (arr[middle] > var) {
				lowerBound = middle - 1;
			}
		}
		return -1; // element doesn't exist
	}	
}

File: lecture005/binarysearch.cpp
//  binary search, upper bound binary search, lower boind binary search, returning the closest element if number not found

#include <iostream>
// #include <climits>		//  library to 
#include <vector>
using namespace std;

void input(vector <int> &arr) {
	for (int i = 0; i < arr.size(); i++) {
		cin>>arr[i];
	}
}

void display(vector <int> &arr) {
	for (int i : arr) {
		cout<<i<<" ";
	}
}

int bin_search(vector <int> &arr, int data) {
	
	int si = 0, ei = arr.size();
	while (si <= ei) {
		int mid = (si + ei)/2;
		if (arr[mid] == data) {
			return mid;
		}
		else if (arr[mid] < data) {
			si = mid + 1;
		}
		else if (arr[mid] > data) {
			ei = mid - 1;
		}
	}

	if (ei == -1) {
		return si;
	}

	else if (si == arr.size()) {
		return ei;
	}

	//  nearest is the digit whose difference with data is smallest
	else if (data - arr[ei] <= arr[si] - data) {
		return ei;		//  if arr[ei] is nearest
	}

	return si;			//  if arr[si] is nearest
}

int l_bin_search(vector<int> &arr, int data) {

	int si = 0, ei = arr.size();
	while (si <= ei)
	{
		int mid = (si + ei) / 2;
		if (arr[mid] == data) {
			if (mid - 1 > -1 && arr[mid - 1] == arr[mid]) {
				ei = mid - 1;
			}
			else {
				return mid;
			}
		}
		else if (arr[mid] < data) {
			si = mid + 1;
		}
		else if (arr[mid] > data) {
			ei = mid - 1;
		}
	}

	if (ei == -1) {
		return si;
	}

	else if (si == arr.size()) {
		return ei;
	}

	//  nearest is the digit whose difference with data is smallest
	else if (data - arr[ei] <= arr[si] - data) {
		return ei; //  if arr[ei] is nearest
	}

	return si; //  if arr[si] is nearest
}

int u_bin_search(vector<int> &arr, int data) {

	int si = 0, ei = arr.size();
	while (si <= ei)
	{
		int mid = (si + ei) / 2;
		if (arr[mid] == data) {
			if (mid + 1 < arr.size() && arr[mid + 1] == arr[mid]) {
				si = mid + 1;
			}
			else {
				return mid;
			}
		}
		else if (arr[mid] < data) {
			si = mid + 1;
		}
		else if (arr[mid] > data) {
			ei = mid - 1;
		}
	}

	if (ei == -1) {
		return si;
	}

	else if (si == arr.size()) {
		return ei;
	}

	//  nearest is the digit whose difference with data is smallest
	else if (data - arr[ei] <= arr[si] - data) {
		return ei; //  if arr[ei] is nearest
	}

	return si; //  if arr[si] is nearest
}

int main(int args, char**argv) {
	
	cout<<"enter size of array: ";
	int size;
	cout<<"size : ";
	cin>>size;
	vector <int> arr(size, 0);
	cout<<"enter the array: ";
	input(arr);
	int data;
	cout<<"data: ";
	cin>>data;
	cout<<bin_search(arr, data)<<endl;
	cout << u_bin_search(arr, data) << endl;
	cout << l_bin_search(arr, data) << endl;
	return 0;
}